### ISSUE TITLE
v1.3.69 failes to save file if Windows notifications are disabled

### ISSUE THREAD

**Issue by asvc** [greenshot/greenshot#182]
**Describe the bug**
File save fails if notification platform is unavailable.

**To Reproduce**
1. Install the latest unstable version, at the time of writing - 1.3.69 - f206046010
2. Capture arbitrary selection, choose "save directly" from the drop-down menu.

**Expected behavior**
1. File saved to the disk in default location

**Screenshots**
![image](https://user-images.githubusercontent.com/300146/79062691-905e4280-7cdb-11ea-9787-d9e65b28b2e0.png)

**Desktop (please complete the following information):**
 - OS: Windows 10 1909 x64 build 18363
 - Browser: N/A
 - Version: 1.3.69 - f206046010 (64 bit); .NET runtime version: 4.0.30319.42000+

**Additional context**
100% reproducible each time "save directly" is chosen.

**Comment by Lakritzator** [comment]
Oh damn, thanks for reporting this.

I did add some check somewhere, maybe this wasn't in the most recent build.
You probably disabled this in a Windows setting?

**Comment by asvc** [comment]
Yes, I have. By the way, the same happens when copying to the clipboard:
```
Software version: 1.3.69 - f206046010 (64 bit)
.NET runtime version: 4.0.30319.42000+
Time: 2020-04-12 21:06:15 +09:30
OS: Windows 10 1909 x64 build 18363
GDI object count: 53
User object count: 47

Exception: System.Exception
Message: The notification platform is unavailable.

The notification platform is unavailable.


Additional Information:
RestrictedDescription : The notification platform is unavailable.

__RestrictedErrorObject : System.Exception+__RestrictedErrorObject
__HasRestrictedLanguageErrorObject : False

Stack:
   at Windows.UI.Notifications.ToastNotificationManager.CreateToastNotifier(String applicationId)
   at GreenshotWin10Plugin.ToastNotificationService.ShowMessage(String message, Int32 timeout, Action onClickAction, Action onClosedAction) in d:\a\1\s\GreenshotWin10Plugin\ToastNotificationService.cs:line 153
   at Greenshot.Helpers.CaptureHelper.SurfaceMessageReceived(Object sender, SurfaceMessageEventArgs eventArgs) in d:\a\1\s\Greenshot\Helpers\CaptureHelper.cs:line 555
   at Greenshot.Drawing.Surface.SendMessageEvent(Object source, SurfaceMessageTyp messageType, String message) in d:\a\1\s\Greenshot\Drawing\Surface.cs:line 1003
   at GreenshotPlugin.Core.AbstractDestination.ProcessExport(ExportInformation exportInformation, ISurface surface) in d:\a\1\s\GreenshotPlugin\Core\AbstractDestination.cs:line 109
   at Greenshot.Destinations.ClipboardDestination.ExportCapture(Boolean manuallyInitiated, ISurface surface, ICaptureDetails captureDetails) in d:\a\1\s\Greenshot\Destinations\ClipboardDestination.cs:line 75
   at GreenshotPlugin.Core.AbstractDestination.<>c__DisplayClass28_0.<ShowPickerMenu>b__4(Object sender, EventArgs e) in d:\a\1\s\GreenshotPlugin\Core\AbstractDestination.cs:line 207
   at System.Windows.Forms.ToolStripItem.RaiseEvent(Object key, EventArgs e)
   at System.Windows.Forms.ToolStripMenuItem.OnClick(EventArgs e)
   at System.Windows.Forms.ToolStripItem.HandleClick(EventArgs e)
   at System.Windows.Forms.ToolStripItem.HandleMouseUp(MouseEventArgs e)
   at System.Windows.Forms.ToolStrip.OnMouseUp(MouseEventArgs mea)
   at System.Windows.Forms.ToolStripDropDown.OnMouseUp(MouseEventArgs mea)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ToolStrip.WndProc(Message& m)
   at System.Windows.Forms.ToolStripDropDown.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

**Comment by Lakritzator** [comment]
If there is a but, it will crash with every notification (although one might discuss if we need one when copying to the clipboard!!)

That's weird, the check for this should be here:
https://github.com/greenshot/greenshot/blob/release/1.3/GreenshotWin10Plugin/ToastNotificationService.cs#L82

If the notifications are not enabled, it should stop and not show.

I will need to check this later today, if my kid doesn't keep me busy, I need some sleep... had a bad night. Without kid I probably already finished with Greenshot 1.3, I would have so much time, but not being able to bring him into the kindergarten, costs me many hours a day.

**Comment by asvc** [comment]
That is a question for someone who has at least tangential knowledge of .NET (e.g. not me), but is there a chance that `toastNotifier` stays uninitialize here?
https://github.com/greenshot/greenshot/blob/5fbd6052390a3f541972956572602e63d634e453/GreenshotWin10Plugin/ToastNotificationService.cs#L82
...and attempts to access `toastNotifier.someRandomCall()` just fail silently instead of raising an exception?

Best of luck with everything you have mentioned in the last paragraph :)

**Comment by Lakritzator** [comment]
I found a small mistake in the Windows 10 Runtime libs I was using, but this was just preventing to run it on 1809 (which is soon obsolete).

Debugging didn't show an issue with the code, so I haven't been able to reproduce it.
Can you update to https://github.com/greenshot/greenshot/releases/tag/v1.3.71 so at least I know if the issue is still there?

If so, you should explain me how you disabled the notifications...

**Comment by asvc** [comment]
v1.3.71 works fine, thanks you. Installed without "Windows 10 plugin" if it matters.

**Comment by asvc** [comment]
> If so, you should explain me how you disabled the notifications...

Here are the applicable HKLM hive changes:
```powershell
reg add "HKLM\Software\Policies\Microsoft\Windows\CloudContent" /v DisableSoftLanding /t REG_DWORD /d 1 /f
reg add "HKLM\Software\Policies\Microsoft\Windows\CloudContent" /v DisableWindowsSpotlightFeatures /t REG_DWORD /d 1 /f
reg add "HKLM\Software\Policies\Microsoft\Windows\CloudContent" /v DisableWindowsConsumerFeatures /t REG_DWORD /d 1 /f
reg add "HKLM\Software\Policies\Microsoft\Windows\DataCollection" /v DoNotShowFeedbackNotifications /t REG_DWORD /d 1 /f
reg add "HKLM\Software\Policies\Microsoft\WindowsInkWorkspace" /v AllowSuggestedAppsInWindowsInkWorkspace /t REG_DWORD /d 0 /f
reg add "HKLM\Software\Microsoft\Windows\CurrentVersion\ImmersiveShell" /v UseActionCenterExperience /t REG_DWORD /d 0 /f
```

Relevant link (content pending update though) -- https://github.com/stoptracking/windows10

**Comment by Lakritzator** [comment]
Thanks for sharing the disabling process, will check it out.

Not installing the Windows 10 plug-in in fact removed the whole Windows 10 notification code which caused the issue, so it might still be present. Without it will fall back to old behavior.

**Comment by asvc** [comment]
Just tested with "Windows 10" plug-in enabled, indeed raises an error:
```log
Software version: 1.3.71 - 3055d42689 (64 bit)
.NET runtime version: 4.0.30319.42000+
Time: 2020-04-13 17:23:15 +09:30
OS: Windows 10 1909 x64 build 18363
GDI object count: 60
User object count: 48

Exception: System.Exception
Message: The notification platform is unavailable.

The notification platform is unavailable.


Additional Information:
RestrictedDescription : The notification platform is unavailable.

__RestrictedErrorObject : System.Exception+__RestrictedErrorObject
__HasRestrictedLanguageErrorObject : False

Stack:
   at Windows.UI.Notifications.ToastNotificationManager.CreateToastNotifier(String applicationId)
   at GreenshotWin10Plugin.ToastNotificationService.ShowMessage(String message, Int32 timeout, Action onClickAction, Action onClosedAction) in d:\a\1\s\GreenshotWin10Plugin\ToastNotificationService.cs:line 153
   at Greenshot.Helpers.CaptureHelper.SurfaceMessageReceived(Object sender, SurfaceMessageEventArgs eventArgs) in d:\a\1\s\Greenshot\Helpers\CaptureHelper.cs:line 563
   at Greenshot.Drawing.Surface.SendMessageEvent(Object source, SurfaceMessageTyp messageType, String message) in d:\a\1\s\Greenshot\Drawing\Surface.cs:line 1003
   at GreenshotPlugin.Core.AbstractDestination.ProcessExport(ExportInformation exportInformation, ISurface surface) in d:\a\1\s\GreenshotPlugin\Core\AbstractDestination.cs:line 99
   at Greenshot.Destinations.FileDestination.ExportCapture(Boolean manuallyInitiated, ISurface surface, ICaptureDetails captureDetails) in d:\a\1\s\Greenshot\Destinations\FileDestination.cs:line 108
   at GreenshotPlugin.Core.AbstractDestination.<>c__DisplayClass28_0.<ShowPickerMenu>b__4(Object sender, EventArgs e) in d:\a\1\s\GreenshotPlugin\Core\AbstractDestination.cs:line 207
   at System.Windows.Forms.ToolStripItem.RaiseEvent(Object key, EventArgs e)
   at System.Windows.Forms.ToolStripMenuItem.OnClick(EventArgs e)
   at System.Windows.Forms.ToolStripItem.HandleClick(EventArgs e)
   at System.Windows.Forms.ToolStripItem.HandleMouseUp(MouseEventArgs e)
   at System.Windows.Forms.ToolStrip.OnMouseUp(MouseEventArgs mea)
   at System.Windows.Forms.ToolStripDropDown.OnMouseUp(MouseEventArgs mea)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ToolStrip.WndProc(Message& m)
   at System.Windows.Forms.ToolStripDropDown.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)

Configuration dump:
```

**Comment by Lakritzator** [comment]
I don't think the mentioned HKLM hive changes are affecting the notification platform...
I can't reproduce it with that, still checking the other mentioned information.

**Comment by asvc** [comment]
Let me try in the fresh VM. Maybe that's a glitch related to the long service history of the current machine, it has survived numerous updates so far.

**Comment by Lakritzator** [comment]
Sorry, I went through all of it, and found nothing that sticks out and screams "this is it".
I cannot try all of that, I have only one system to try with and don't want to spent hours of trying every single setting or all and trying to reverse them, to be able to find what is causing this.

I really would like to know what settings is causing this, and how to detect this.
From my opinion, if people start going through so many modifications, they are asking for problems 😀 

For now I've added two checks:
Check if the feature if generally available, don't know if this helps.
And catching the error, ignoring it, which is not my favorite way but it should make it usable again.
There will be no notification fallback than, Greenshot will be completely silent.

P.S.
I barely have enough memory to run Windows & Visual Studio on my system 😢
I am collecting money to buy a new system, I should have given my system more RAM back than.

**Comment by asvc** [comment]
For the sake of completeness, or in case someone else stumbles upon this issue later:

First start after the installation, fresh Windows 10 VM has greeted me with this:
```
Software version: 1.3.71 - 3055d42689 (64 bit)
.NET runtime version: 4.0.30319.42000+
Time: 2020-04-14 20:00:38 +09:30
OS: Windows 10 1903 x64 build 18362
GDI object count: 25
User object count: 23

Exception: System.Exception
Message: Element not found. (Exception from HRESULT: 0x80070490)

Stack:
   at Windows.UI.Notifications.ToastNotifier.get_Setting()
   at GreenshotWin10Plugin.ToastNotificationService.ShowMessage(String message, Int32 timeout, Action onClickAction, Action onClosedAction) in d:\a\1\s\GreenshotWin10Plugin\ToastNotificationService.cs:line 83
   at Greenshot.MainForm.HandleDataTransport(CopyDataTransport dataTransport) in d:\a\1\s\Greenshot\Forms\MainForm.cs:line 449
   at Greenshot.MainForm..ctor(CopyDataTransport dataTransport) in d:\a\1\s\Greenshot\Forms\MainForm.cs:line 422
   at Greenshot.MainForm.Start(String[] arguments) in d:\a\1\s\Greenshot\Forms\MainForm.cs:line 273

Configuration dump:
```

Screenshot (open in the new tab, it's large):
![2020-04-14 20_01_48-Windows Sandbox](https://user-images.githubusercontent.com/300146/79215884-6e0b2700-7e8b-11ea-9e7a-c51b7f29f43b.png)

Restarting the application worked fine, but an attempt to take a screenshot has ended-up with the following error:
```
Software version: 1.3.71 - 3055d42689 (64 bit)
.NET runtime version: 4.0.30319.42000+
Time: 2020-04-14 20:05:01 +09:30
OS: Windows 10 1903 x64 build 18362
GDI object count: 36
User object count: 37

Exception: System.Runtime.InteropServices.COMException
Message: Error HRESULT E_FAIL has been returned from a call to a COM component.

ErrorCode: 0x80004005

Stack:
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GreenshotWin10Plugin.Win10OcrProvider.<DoOcrAsync>d__4.MoveNext() in d:\a\1\s\GreenshotWin10Plugin\Win10OcrProvider.cs:line 108
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GreenshotWin10Plugin.Win10OcrProvider.<DoOcrAsync>d__2.MoveNext() in d:\a\1\s\GreenshotWin10Plugin\Win10OcrProvider.cs:line 69
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GreenshotWin10Plugin.Processors.Win10OcrProcessor.<>c__DisplayClass4_0.<<ProcessCapture>b__0>d.MoveNext() in d:\a\1\s\GreenshotWin10Plugin\Processors\Win10OcrProcessor.cs:line 44

Configuration dump:
```

And another large screenshot:
![2020-04-14 20_07_55-Windows Sandbox](https://user-images.githubusercontent.com/300146/79216063-b4f91c80-7e8b-11ea-941f-919b46923d2e.png)

If this bug is isolated & non-reproducible (which is quite likely; long history of self-updating Windows 10), then let's close this issue?


**P.S.** Feel free to shoot me an e-mail re. RAM/etc at andrew@savchenko.net

**Comment by Lakritzator** [comment]
What kind of VM image are you using? This doesn't really make sense...

I'm going to try to reproduce it on a different PC, but this takes a bit of time, they broke an internet cable around my place, probably >100.000 of people don't have internet and the mobile internet is crawling...  😢 

P.S.
You can't help me with RAM, I bought a Surface Pro 2017, with only 8 GB. The SSD started to be crap, slowest every, shortly after the guarantee. I'm waiting for the 3 years to pass, until than I'm developing on my working PC Laptop which is highly restricted (no admin rights).

**Comment by asvc** [comment]
Windows sandbox with default settings. Can spin-up the fully-fledged Hyper-V instance if you'd like.

Regarding this particular issue - as it is not affecting many users (so far - one), is it worth your time?

P.P.S. No worries, just feel free to message me if this becomes a show-stopper. Stranger things have happened than magical solution to one unfortunate MS Surface.

**Comment by Lakritzator** [comment]
I'll try it out with the sandbox, it should somewhat run... slowly though.

I think this is worth my time, as you are one of the first persons to run the new 1.3 code. The notifications for Windows 10 are new, building a base for later functionality.

**Comment by asvc** [comment]
I have tried latest stable, selected "compact installation" and it has worked just fine.
The only notable difference was this window staying up for much longer:
![2020-04-14 22_04_58-Windows Sandbox](https://user-images.githubusercontent.com/300146/79225559-10330b00-7e9c-11ea-84a8-3d870e2b9e9f.png)

Perhaps it is related to some pre-compilation .NET does after the installation stage?

Bonus screenshot showing the working notification:
![2020-04-14 22_07_09-Windows Sandbox](https://user-images.githubusercontent.com/300146/79225738-54bea680-7e9c-11ea-9990-35ffb1d22a98.png)

**Comment by Lakritzator** [comment]
With the latest stable, the Windows 10 notifications were not included.
These are the old notifications, and offer a lot less functionality.

The new notification can be recognized as the Greenshot icon is also in there:
![image](https://user-images.githubusercontent.com/708125/79229093-c5e16800-7e62-11ea-9312-532cc5745bf9.png)

Depending on the destination, we can add additional buttons like after uploading it to a cloud service a button could open the screenshot in a web browser.

About that it took longer to install, in the newer version I removed the "NGEN" (native generation) step as it doesn't bring much with .NET >= 4.7

**Comment by Lakritzator** [comment]
Ah, I can reproduce the Sandbox issue you had!

Making it reproducible is the first step for fixing things 😄 

```
Software version: 1.3.71 - 3055d42689 (64 bit)
.NET runtime version: 4.0.30319.42000+
Time: 2020-04-14 15:23:21 +02:00
OS: Windows 10 2004 x64 build 19041
GDI object count: 28
User object count: 24

Exception: System.Exception
Message: Element not found. (Exception from HRESULT: 0x80070490)

Stack:
   at Windows.UI.Notifications.ToastNotifier.get_Setting()
   at GreenshotWin10Plugin.ToastNotificationService.ShowMessage(String message, Int32 timeout, Action onClickAction, Action onClosedAction) in d:\a\1\s\GreenshotWin10Plugin\ToastNotificationService.cs:line 83
   at Greenshot.MainForm.HandleDataTransport(CopyDataTransport dataTransport) in d:\a\1\s\Greenshot\Forms\MainForm.cs:line 449
   at Greenshot.MainForm..ctor(CopyDataTransport dataTransport) in d:\a\1\s\Greenshot\Forms\MainForm.cs:line 422
   at Greenshot.MainForm.Start(String[] arguments) in d:\a\1\s\Greenshot\Forms\MainForm.cs:line 273
```

**Comment by Lakritzator** [comment]
FYI: Internet is working again 🥳 
But now I need to spend a bit of time with my family... I'll try some more tonight.

**Comment by Lakritzator** [comment]
I found the issue with the "Element not found" error, Windows APIs are never boring:
https://www.rudyhuyn.com/blog/2018/02/10/toastnotifier-and-settings-careful-with-non-uwp-applications/

I already tried the fix, I still need to commit & push it, but I need my sleep now.

**Comment by Lakritzator** [comment]
Solved with 1.3.75, thanks for reporting this and helping me solve it.
